src/lib/util/opresolv.cpp

```diff
@@ -19,546 +19,465 @@
+bool option_resolution::entry::bump_lower()
 {
-	option_resolution::error err;
-	range ranges[256];
-	int i;
-
-	err = list_ranges(specification, option_char, ranges, ARRAY_LENGTH(ranges));
-	if (err != error::SUCCESS)
-		return err;
+	auto old_value = value_int();
+	auto current_range = find_in_ranges(old_value);
+	assert(current_range != m_ranges.end());
 
-	for (i = 0; (ranges[i].min >= 0) && (ranges[i].max >= 0); i++)
+	int new_value;
+	if (old_value > current_range->min)
+	{
+		// decrement within current range
+		new_value = old_value - 1;
+	}
+	else if (current_range != m_ranges.begin())
+	{
+		// go to the top of the previous range
+		new_value = (current_range - 1)->max;
+	}
+	else
 	{
-		if ((ranges[i].min <= val) && (ranges[i].max >= val))
-			return error::SUCCESS;
+		// at the minimum; don't bump
+		new_value = old_value;
 	}
-	return error::PARAMOUTOFRANGE;
+
+	set_value(new_value);
+	return new_value != old_value;
 }
 
 
 // -------------------------------------------------
-//  contains
+//	entry::bump_higher
 // -------------------------------------------------
 
-bool option_resolution::contains(const char *specification, int option_char)
+bool option_resolution::entry::bump_higher()
 {
-	return strchr(specification, option_char) != nullptr;
+	auto old_value = value_int();
+	auto current_range = find_in_ranges(old_value);
+	assert(current_range != m_ranges.end());
+
+	int new_value;
+	if (old_value < current_range->max)
+	{
+		// increment within current range
+		new_value = old_value + 1;
+	}
+	else if (current_range != (m_ranges.end() - 1))
+	{
+		// go to the bottom of the next range
+		new_value = (current_range + 1)->min;
+	}
+	else
+	{
+		// at the minimum; don't bump
+		new_value = old_value;
+	}
+
+	set_value(new_value);
+	return new_value != old_value;
 }
 
 
 // -------------------------------------------------
-//  error_string
+//	entry::find_in_ranges
 // -------------------------------------------------
 
-const char *option_resolution::error_string(option_resolution::error err)
+option_resolution::entry::rangelist::const_iterator option_resolution::entry::find_in_ranges(int value) const
 {
-	switch (err)
-	{
-	case error::SUCCESS:                return "The operation completed successfully";
-	case error::OUTOFMEMORY:            return "Out of memory";
-	case error::PARAMOUTOFRANGE:        return "Parameter out of range";
-	case error::PARAMNOTSPECIFIED:      return "Parameter not specified";
-	case error::PARAMNOTFOUND:          return "Unknown parameter";
-	case error::PARAMALREADYSPECIFIED:  return "Parameter specified multiple times";
-	case error::BADPARAM:               return "Invalid parameter";
-	case error::SYNTAX:                 return "Syntax error";
-	case error::INTERNAL:               return "Internal error";
-	}
-	return nullptr;
+	return std::find_if(
+		m_ranges.begin(),
+		m_ranges.end(),
+		[&](const auto &r) { return r.min <= value && value <= r.max; });
```

Comment:
Please don't use auto-capture, make it explicit what the lambda needs to capture.